docs(lez): Extensions section with admin-authority and extension-library guides (#355)
status-im/ift-docs: docs/lez/extensions/build-a-spel-extension-library.md

When two extensions embed into the same consumer account at distinct offsets, the framework merges the duplicated account into one transaction account (listed once in the IDL with unioned constraints, cloned into each position of the call) and your instruction must emit exactly one post-state per unique account id. Same account at the same offset is a compile error.

## Consumer integration

A consumer adds your extension to their `Cargo.toml`:

```toml
[dependencies]
my-extension = { git = "https://github.com/you/my-extension" }

> spel-framework = { git = "https://github.com/logos-co/spel" }

Author: Fixed in 57e4d16. The consumer snippet pins the fork rev, with the same branch-vs-rev caveat as the admin page and a note to swap to logos-co once https://github.com/logos-co/spel/pull/255 lands.

## Multiple extensions on one program

Consumers can stack extensions without coordination between library authors:

```rust
#[lez_program]
#[admin_authority]
#[my_extension]
mod my_program { ... }
```

Each extension is discovered independently by its own `extension_attr`. Each contributes its own instructions to the dispatcher, and each gate attribute re-expands on its own gated handlers without touching the others.

Marker order is the cross-extension ABI: the first marker's instructions and injected parameters come first in the dispatcher, the IDL, and the account order. When two extensions inject the same parameter name with identical constraints they share one account, conflicting constraints are a compile error naming both extensions. Duplicate instruction names between extensions (or an extension and a consumer function) are a compile error naming both sources. Two extensions can even embed into the same consumer account at distinct offsets, see embedded mode above.

## Verifying your extension

Build a small sample program that consumes your extension. Then:

```bash
spel generate-idl path/to/sample/src/main.rs
```

> The IDL should contain your extension's instructions alongside the consumer's own. A marker that matches no discoverable extension is a hard compile error naming the marker, regardless of why it did not match, so a broken setup refuses loudly instead of building a program silently missing its extension surface. When you hit that error, the most common causes are:

Author: Fixed in 57e4d16, including the hint to check the framework pin before the metadata when debugging a missing surface.

```
my-extension/
├── my-extension/             # runtime library: types, instruction fns, metadata
│   ├── Cargo.toml
│   └── src/lib.rs
├── my-extension-macros/      # proc-macro sub-crate: marker + gate attributes
│   ├── Cargo.toml
│   └── src/lib.rs
└── my-extension-sample/      # reference SPEL program (optional but recommended)
    └── ...
```

The split exists because Rust requires proc-macro attributes to live in a `proc-macro = true` crate that cannot also export non-macro items. The runtime library re-exports the macros so consumers only declare one dependency.

## The discovery metadata

In `my-extension/Cargo.toml`:

```toml
[package]
name = "my-extension"
version = "0.1.0"
edition = "2021"

[package.metadata.spel]
extension_attr = "my_extension"

[dependencies]
borsh = { version = "1", features = ["derive"] }
spel-framework = { git = "https://github.com/mmlado/spel", rev = "f7aa464b2c6c72ef513a25ede16584bca85b722f" }
my-extension-macros = { path = "../my-extension-macros" }
```

- `extension_attr` is the attribute name consumers put on their `#[lez_program]` module to opt in. By convention, match it to your crate name (with `_` not `-`).

> Per-instruction gate attributes your library defines (e.g. `#[require_admin]` from `admin-authority`) need no metadata for the check itself: they are ordinary proc-macros that re-expand on the emitted handler and consume themselves, so the framework leaves them alone. If your gate needs specific account parameters on every gated instruction, you can declare those in an optional inject block so consumers do not have to write them out (see the gate attribute section below).

Reviewer: `Google.Latin` is error-level and fires three times on this branch. Same fix at `build-a-spel-extension-library.md` line 284 (`E.g.` → `For example,`) and `freeze-authority.md` line 227 (`e.g.`).

```suggestion
Per-instruction gate attributes your library defines (for example, `#[require_admin]` from `admin-authority`) need no metadata for the check itself: they are ordinary proc-macros that re-expand on the emitted handler and consume themselves, so the framework leaves them alone. If your gate needs specific account parameters on every gated instruction, you can declare those in an optional inject block so consumers do not have to write them out (see the gate attribute section below).
```

```rust
#[require_my_gate]   // runs first — sees params with #[account(...)] intact
#[instruction]       // shim runs second — strips #[account(...)] for rustc
pub fn gated_op(/* ... */) -> SpelResult { /* ... */ }
```

Rust expands attribute macros top-down. The library's `#[instruction]` shim strips `#[account(...)]` from parameters. If `#[require_my_gate]` is placed below `#[instruction]`, it runs after the strip, no PDA or signer parameters are left for its shape check, and it emits a confusing error.

The rule only applies inside libraries that re-export the shim (like the one shown in this guide). Consumer code uses SPEL's no-op `#[instruction]` from the prelude, which doesn't strip anything; order doesn't matter there.

## Composing with another extension (hard dep)

Some extensions naturally build on others. `freeze-authority` depends on `admin-authority` — its freeze-authority slot is governed by admin signatures. When your extension does this:

> 1. **Declare a normal Cargo path dep** on the other extension in your `Cargo.toml`. Consumers get both extensions in their dep graph automatically.

Reviewer: "path dep" is too narrow, and it contradicts the example the numbered list itself cites. `freeze-authority` composes with `admin-authority` over a **git** dependency, not a path one:

```toml
# freeze-authority/Cargo.toml @ v0.1.0
admin-authority = { git = "https://github.com/mmlado/spel-admin-authority.git", tag = "v0.1.0" }
```

Line 301 already says "Path, git, and registry dependencies are all discoverable", so this step reads as a stricter requirement than the mechanism actually imposes. (Line 287's "deduplicates path-dep dirs" is fine as-is — that one really is about path deps.)

```suggestion
1. **Declare a normal Cargo dependency** on the other extension in your `Cargo.toml`, path or git — `freeze-authority` uses a git dependency on `admin-authority` pinned to its `v0.1.0` tag. Consumers get both extensions in their dep graph automatically.
```

2. **Add both markers to the consumer's mod.** Consumers write `#[admin_authority] #[my_extension]` on their `#[lez_program]` mod. Each marker triggers its own discovery.
3. **Import the gate attributes you compose with.** E.g. `use admin_authority::require_admin;` in your library source, then `#[require_admin]` on instructions that should require admin sig (like an initialization that creates your config PDA).
4. **List the other extension's exempt-while-wrapped instructions** in your `wrap_instructions.exempt` if applicable. freeze-authority lists admin-authority's three management instructions so they stay callable while the program is frozen.

The framework deduplicates path-dep dirs, so admin-authority is scanned once even if both your extension and the consumer name it as a path dep.

## Consumer integration

A consumer adds your extension to their `Cargo.toml`:

```toml
[dependencies]
my-extension = { git = "https://github.com/you/my-extension" }

> spel-framework = { git = "https://github.com/mmlado/spel", rev = "f7aa464b2c6c72ef513a25ede16584bca85b722f" }

Reviewer: **This consumer dependency block does not compile.** I built the toy extension from this guide end to end — the two crates in [Layout](#layout), the metadata, the marker macro, the `#[instruction]` shim — then wrote the sample consumer with exactly these two dependencies, and got four hard errors before any of my own code was reached:

```
error[E0433]: failed to resolve: use of unresolved module or unlinked crate `serde`
 --> src/main.rs:3:1
  |
3 | #[lez_program]
  | ^^^^^^^^^^^^^^ use of unresolved module or unlinked crate `serde`

error[E0433]: failed to resolve: use of unresolved module or unlinked crate `nssa_core`   (x3)
error[E0282]: type annotations needed
```

`#[lez_program]` expands to code that names `nssa_core` and `serde` directly, so the consumer must declare both even though their own source never mentions either. `admin-authority.md` gets this right (its block carries all five, with "All five are needed"); this page's block is missing two, which bites library authors at exactly the point the guide tells them to build a sample to verify their work.

Adding the two makes it build, and `spel generate-idl` then merges the extension's instruction in as the guide promises:

```
    Finished `dev` profile [unoptimized + debuginfo] target(s)
$ spel generate-idl src/main.rs | jq -r '.instructions[].name'
my_user_instr
extension_action
```

```suggestion
my-extension = { git = "https://github.com/you/my-extension" }
spel-framework = { git = "https://github.com/mmlado/spel", rev = "f7aa464b2c6c72ef513a25ede16584bca85b722f" }
nssa_core = { git = "https://github.com/logos-blockchain/logos-execution-zone.git", tag = "v0.2.0", package = "lee_core" }
serde = { version = "1", features = ["derive"] }
```

}
```

Three things to note:

- `extern crate self as my_extension;`, lets the library reference its own types via the absolute path `::my_extension::MyState`. The framework emits cross-crate calls into the consumer's binary using that path, so the path needs to resolve both in the library's own compile and at the consumer's compile.
- `pub use my_extension_macros::{instruction, my_extension};`, re-exports the marker attribute and the no-op `#[instruction]` shim so consumers (and the library's own `lib.rs`) can use them without importing the macros crate directly.
- `#[account(...)]` attributes on parameters, these are framework helper attributes that describe PDA seeds, signer requirements, etc. The library's own `#[instruction]` shim strips them at the library's compile so rustc accepts the source; the framework reads them during the path-dependency scan.
- Name the state parameter after the inject role you will declare for it (`my_state` here). Injection reuse, wrap stamping, and embedded retargeting resolve your accounts by role name, a differently named parameter breaks embedded mode with an argument-count error at the consumer's compile.

> - When you write the real body, post-states are the inner `account` values (`vec![my_state.account, caller.account]`), with an `(account, AutoClaim)` tuple for accounts the instruction claims. Consumer handlers return the `AccountWithMetadata` wrappers, library handlers do not. The reference samples show both patterns.

Reviewer: Following this literally produces code that does not compile, and it is the first thing an author hits when they replace the `todo!()` with a real body.

The sentence reads as though a single vec can hold bare `account` values *and* an `(account, AutoClaim)` tuple for the claimed ones. It cannot — the vec is homogeneous. I tested all three shapes against `my-extension` built from this guide:

```rust
// works
vec![my_state.account, caller.account]

// works
vec![(my_state.account, AutoClaim::None), (caller.account, AutoClaim::None)]

// what the sentence reads like — does NOT compile
vec![my_state.account, (caller.account, AutoClaim::None)]
```

The last one fails with:

```
error[E0308]: mismatched types
  --> src/lib.rs:28:51
   |
28 |     Ok(SpelOutput::execute(vec![my_state.account, (caller.account, AutoClaim::None)], vec![]))
   |                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Account`, found `(Account, AutoClaim)`
```

`admin-authority`'s own source is the proof of the real rule: `admin_transfer` returns plain `vec![admin_config.account, caller.account, new_account.account]`, but `admin_initialize` — which claims the config PDA — makes *every* entry a tuple, including the unclaimed caller:

```rust
vec![
    (config.account, AutoClaim::Claimed(Claim::Pda(PdaSeed::new(seed_from_str("admin_config"))))),
    (caller.account, AutoClaim::None),
]
```

So the rule is all-or-nothing per instruction, not per account.

```suggestion
- When you write the real body, post-states are the inner `account` values (`vec![my_state.account, caller.account]`). If the instruction claims any account, every entry in that vec becomes an `(account, AutoClaim)` tuple instead, with `AutoClaim::None` for the ones it does not claim, the way `admin-authority`'s `admin_initialize` returns `vec![(config.account, AutoClaim::Claimed(..)), (caller.account, AutoClaim::None)]`. The vec is homogeneous either way, mixing a bare `account` with a tuple fails to compile with `expected Account, found (Account, AutoClaim)`. Consumer handlers return the `AccountWithMetadata` wrappers, library handlers do not. The reference samples show both patterns.
```

Author: Applied with your wording. Thank you for getting far enough to write real bodies.

#[proc_macro_attribute]
pub fn require_my_gate(_attr: TokenStream, item: TokenStream) -> TokenStream {
    let mut func: ItemFn = match syn::parse(item) {
        Ok(f) => f,
        Err(e) => return e.to_compile_error().into(),
    };

    // Prepend the runtime check. It references params by their conventional
    // names; accept attribute args to let consumers override the names.
    let prologue: syn::Stmt = parse_quote! {{
        let __state = ::my_extension::MyState::from_account(&my_state)?;
        __state.assert_allowed(&caller)?;
    }};
    func.block.stmts.insert(0, prologue);

    quote::quote!(#func).into()
}
```

The `from_account` and `assert_allowed` helpers the prologue calls are yours to write on `MyState`, the framework provides neither. Re-export the gate from the runtime library next to the marker (`pub use my_extension_macros::require_my_gate;`) so consumers import everything from one crate.

Never read or strip `#[account(...)]` attributes in a gate macro. That attribute belongs to the framework, which reads it for validation and the IDL. Your gate should only reference parameter names, taken from its own attribute args with sensible defaults.

> **The kwarg contract.** Your gate's attribute keys must be exactly the inject-account names you declare in metadata (`#[require_my_gate(my_state = their_cfg, caller = owner)]`). The framework's auto-wrap and gate stamping emit every kwarg with the resolved parameter name, so your macro receives the framework's naming decisions instead of guessing from convention. Ship an alignment self-test so the two cannot drift: read your own metadata with `spel_framework_core::extension::read_inject_specs(Path::new(env!("CARGO_MANIFEST_DIR")))` and assert the declared account names equal the kwarg set a probe function hands your gate. A name the macro rejects fails the probe's compile, a metadata rename fails the runtime assert.

Reviewer: **The alignment self-test cannot be written against the `Cargo.toml` this guide gives — `spel_framework_core` does not resolve.**

I built the toy extension from this guide end to end (both crates, the metadata, the marker, the `#[require_my_gate]` gate, the inject block), confirmed the injection works — a consumer instruction carrying only its own account comes out of the IDL as `["my_state","caller","user_state"]`, exactly as L206 promises — and then tried to ship the self-test this paragraph asks for. It does not compile:

```
error[E0433]: failed to resolve: use of unresolved module or unlinked crate `spel_framework_core`
 --> tests/alignment.rs:6:17
```

Two separate reasons, and the second one is the non-obvious half:

1. The library manifest in [The discovery metadata](#the-discovery-metadata) declares `spel-framework`, not `spel-framework-core`, and the core crate is not re-exported under a nameable path. `spel-framework/src/lib.rs` does `pub use spel_framework_core::*`, but that does not give you `spel_framework::extension`:

   ```
   error[E0433]: failed to resolve: could not find `extension` in `spel_framework`
   ```

2. Adding the dev-dependency plainly still fails, because `extension` is behind a **non-default** feature — `spel-framework-core/src/lib.rs:19` is `#[cfg(feature = "idl-gen")] pub mod extension;`, and `[features] default = ["decode"]`:

   ```
   error[E0433]: failed to resolve: could not find `extension` in `spel_framework_core`
   ```

With `features = ["idl-gen"]` the test the paragraph describes passes on the first run:

```
$ cargo test --test alignment
running 1 test
test result: ok. 1 passed; 0 failed
```

For the record, the signature is `read_inject_specs(crate_dir: &Path) -> Result<Vec<InjectSpec>, String>` (`spel-framework-core/src/extension.rs:504`), so the test needs an `.expect(..)` — worth a word, since the sentence reads as though it hands back the specs directly.

````suggestion
**The kwarg contract.** Your gate's attribute keys must be exactly the inject-account names you declare in metadata (`#[require_my_gate(my_state = their_cfg, caller = owner)]`). The framework's auto-wrap and gate stamping emit every kwarg with the resolved parameter name, so your macro receives the framework's naming decisions instead of guessing from convention. Ship an alignment self-test so the two cannot drift: read your own metadata with `spel_framework_core::extension::read_inject_specs(Path::new(env!("CARGO_MANIFEST_DIR")))`, which returns `Result<Vec<InjectSpec>, String>`, and assert the declared account names equal the kwarg set a probe function hands your gate. A name the macro rejects fails the probe's compile, a metadata rename fails the runtime assert. The `extension` module sits behind `spel-framework-core`'s non-default `idl-gen` feature and is not reachable through `spel-framework`, so the test needs its own dev-dependency:

```toml
[dev-dependencies]
spel-framework-core = { git = "https://github.com/mmlado/spel", rev = "f7aa464b2c6c72ef513a25ede16584bca85b722f", features = ["idl-gen"] }
```
````

---
_Generated by [Claude Code](https://claude.ai/code)_

Author: Applied: the Result return, the dev-dependency block, and the non-default idl-gen feature note.
